src/Fabulous/Builders.fs

```diff
@@ -2,139 +2,229 @@ namespace Fabulous
+        member x.Compile() : Widget =
+            let struct (scalarAttributes, widgetAttributes, widgetCollectionAttributes) = x.Attributes
+
+            {
+                Key = x.Key
 #if DEBUG
-            DebugName = $"{typeof<'marker>.Name}<{typeof<'msg>.Name}>"
+                DebugName = $"{typeof<'marker>.Name}<{typeof<'msg>.Name}>"
 #endif
-            ScalarAttributes = scalarAttributes
-            WidgetAttributes = widgetAttributes
-            WidgetCollectionAttributes = widgetCollectionAttributes
-        }
-
-    [<EditorBrowsable(EditorBrowsableState.Never)>]
-    member _.AddScalar(attr: ScalarAttribute) =
-        let struct (scalarAttributes, widgetAttributes, widgetCollectionAttributes) = attributes
-        let attribs = scalarAttributes
-        let attribs2 = Array.zeroCreate(attribs.Length + 1)
-        Array.blit attribs 0 attribs2 0 attribs.Length
-        attribs2.[attribs.Length] <- attr
-
-        WidgetBuilder<'msg, 'marker>(key, struct (attribs2, widgetAttributes, widgetCollectionAttributes))
-
-    [<EditorBrowsable(EditorBrowsableState.Never)>]
-    member _.AddWidget(attr: WidgetAttribute) =
-        let struct (scalarAttributes, widgetAttributes, widgetCollectionAttributes) = attributes
-        let attribs = widgetAttributes
-        let attribs2 = Array.zeroCreate(attribs.Length + 1)
-        Array.blit attribs 0 attribs2 0 attribs.Length
-        attribs2.[attribs.Length] <- attr
-
-        WidgetBuilder<'msg, 'marker>(key, struct (scalarAttributes, attribs2, widgetCollectionAttributes))
-
-    [<EditorBrowsable(EditorBrowsableState.Never)>]
-    member _.AddWidgetCollection(attr: WidgetCollectionAttribute) =
-        let struct (scalarAttributes, widgetAttributes, widgetCollectionAttributes) = attributes
-        let attribs = widgetCollectionAttributes
-        let attribs2 = Array.zeroCreate(attribs.Length + 1)
-        Array.blit attribs 0 attribs2 0 attribs.Length
-        attribs2.[attribs.Length] <- attr
-
-        WidgetBuilder<'msg, 'marker>(key, struct (scalarAttributes, widgetAttributes, attribs2))
-
-    [<EditorBrowsable(EditorBrowsableState.Never)>]
-    member x.AddScalars(attrs: ScalarAttribute []) =
-        if attrs.Length = 0 then
-            x
-        else
-            let struct (scalarAttributes, widgetAttributes, widgetCollectionAttributes) = attributes
-            let attribs = scalarAttributes
-
-            let attribs2 =
-                Array.zeroCreate(attribs.Length + attrs.Length)
-
-            Array.blit attribs 0 attribs2 0 attribs.Length
-            Array.blit attrs 0 attribs2 attribs.Length attrs.Length
-
-            WidgetBuilder<'msg, 'marker>(key, struct (attribs2, widgetAttributes, widgetCollectionAttributes))
+                ScalarAttributes =
+                    match StackList.length &scalarAttributes with
+                    | 0us -> None
+                    | _ -> Some(Array.sortInPlace(fun a -> a.Key) (StackList.toArray &scalarAttributes))
```

Comment:
Same for this one. Widget being struct means we can remove allocation of `Some` by using `ValueSome`

Reply:
My understanding is that `Some` for reference types does not allocate extra, and is represented as `null` at runtime.

Reply:
In that particular case it is `ScalarAttribute[]` which is reference nullable type. let me see decompiled output :)

Reply:
Hm, I think you are right!

Option actually a reference type allocated on the heap.
https://www.decompiler.com/jar/892c5a77b8234d2ebeec4873cb37b00d/Fabulous.dll/Fabulous/WidgetBuilder.cs

```csharp
scalarAttributes = FSharpOption<ScalarAttribute[]>.Some(array);
```

And 

<img width="644" alt="image" src="https://user-images.githubusercontent.com/6633216/148637424-fe3a6125-03ec-481c-b124-fee397a9b34a.png">

Fascinating!